src/puppetlabs/dujour/version_check.clj

```diff
@@ -0,0 +1,116 @@
+(ns puppetlabs.dujour.version-check
+  (:require [clojure.tools.logging :as log]
+            [schema.core :as schema]
+            [ring.util.codec :as ring-codec]
+            [puppetlabs.http.client.sync :as client]
+            [cheshire.core :as json]
+            [trptcolin.versioneer.core :as version]
+            [slingshot.slingshot :as sling]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Constants
+
+(def default-group-id "puppetlabs.packages")
+(def default-update-server-url "http://updates.puppetlabs.com")
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Schemas
+
+(def ProductCoords
+  {:group-id schema/Str
+   :artifact-id schema/Str})
+
+(def ProductName
+  (schema/conditional
+    #(string? %) schema/Str
+    #(map? %) ProductCoords))
+
+(def UpdateInfo
+  {:version schema/Str
+   :newer   schema/Bool
+   :link    schema/Str
+   :product schema/Str
+   (schema/optional-key :message) schema/Str})
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Private
+
+(defn schema-match?
+  [schema obj]
+  (nil? (schema/check schema obj)))
+
+(schema/defn get-coords :- ProductCoords
+  [product-name :- ProductName]
+  (condp schema-match? product-name
+    schema/Str {:group-id default-group-id
+                :artifact-id product-name}
+    ProductCoords product-name))
+
+(schema/defn version*
+  "Get the version number of this installation."
+  [group-id artifact-id]
+  {:post [(string? %)]}
+  (version/get-version group-id artifact-id))
+
+(def version
+  "Get the version number of this installation."
+  (memoize version*))
+
+(schema/defn ^:always-validate update-info :- (schema/maybe UpdateInfo)
+  "Make a request to the puppetlabs server to determine the latest available
+  version. Returns the JSON object received from the server, which
+  is expected to be a map containing keys `:version`, `:newer`, and `:link`.
+  Returns `nil` if the request does not succeed for some reason."
+  [product-name :- ProductName
+   update-server-url :- schema/Str]
+  (let [{:keys [group-id artifact-id]} (get-coords product-name)
+        current-version (version group-id artifact-id)
+        version-data {:version current-version}
+        query-string (ring-codec/form-encode version-data)
+        url (format "%s?product=%s&group=%s&%s" update-server-url artifact-id group-id query-string)
+        {:keys [status body] :as resp} (client/get url
```

Comment:
I don't see much information in the client request beyond the version information.  Would it be a good idea to include some additional information about the client so we can build, measure, learn faster?

I'm thinking something trivial as a straw-man for how we might want to extend this in the future.  For example, maybe a client request header that has a value for the configured JVM heap size.  How would it look for us to support arbitrary pieces of information being passed back?  Should it be part of the request body?  Might it need to be encrypted?  Should we not answer any of these now, because we're confident it'll be possible to do so when the time comes?